Track full-path htlc-minimum-msat while routing

TheBlueMatt opened the pull request:
To review new changes to #646, I wrote an extension of the router fuzzer to test each route against the graph available. This found one non-regression which is fixed here, as well as the fuzzer extension and a fix to make the router fuzz target deterministic.

It needs a) a test, b) a long discussion in `get_route` about how our algorithm ultimately differs from Dijkstra's/A*, why that's ok, and some edge cases where other algorithms may be more optimal.

TheBlueMatt commented:
Pushed a few more fix commits (including some regressions from MPP) including tests for each bug. There may be one more, but these all stand.

naumenkogs commented on lightning/src/routing/router.rs:
```diff
@@ -901,13 +908,23 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
 			break 'paths_collection;
 		}
+		// Further, if this was our first walk of the graph, and we weren't limited by an
+		// htlc_minim_msat, return immediately because this path should suffice. If we were limited
```
typo: htlc_minim_msat

naumenkogs commented on lightning/src/routing/router.rs:
```diff
@@ -901,13 +908,23 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
+		// htlc_minim_msat, return immediately because this path should suffice. If we were limited
+		// by an htlc_minimum_msat value, find another path with a higher value, potentially
+		// allowing us to pay fees to meet the htlc_minimum while still keeping a lower total fee.
+		if already_collected_value_msat == final_value_msat && payment_paths.len() == 1 {
```
nit: this is mutually exclusive with the previous if because of `found_new_path`, I think it would be easier to understand their relation if merged via `else`

naumenkogs commented on lightning/src/routing/router.rs:
```diff
@@ -3371,6 +3388,77 @@ mod tests {
+		});
+
+		{
+			// Now, attempt to route 125 sats (just a bit below the capacity of 3 channels).
```
this comment says 125, but you send 90 :)

naumenkogs replied:
And the next line is outdated too.

naumenkogs commented on lightning/src/routing/router.rs:
```diff
@@ -778,17 +790,17 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
 				// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
 				// really sucks, cause we're gonna need that eventually.
-				let last_hop_htlc_minimum_msat: u64 = match hop.htlc_minimum_msat {
+				let last_path_htlc_minimum_msat: u64 = match hop.htlc_minimum_msat {
```
I still don't understand this change.

ariard replied:
You mean the variable renaming or the checking along each hop path that the amount transferred is always above worst-hop's-htlc_minimum_msat-along-the-path ? Where such constraint only apply to hops after such value (i.e closer to payer).

naumenkogs replied:
I don't understand the renaming.

TheBlueMatt replied:
Honestly, I'm not sure either. I reverted the change.

naumenkogs commented on lightning/src/routing/router.rs:
```diff
@@ -3459,6 +3471,66 @@ mod tests {
+		});
+
+		{
+			// Now, attempt to route 125 sats (just a bit below the capacity of 3 channels).
```
outdated

naumenkogs commented:
Minor comments, ACK 68b62c0116206debe268eca898d48c786d6f16d4

ariard commented on lightning/src/routing/router.rs:
```diff
@@ -861,19 +910,29 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
+		// by an htlc_minimum_msat value, find another path with a higher value, potentially
+		// allowing us to pay fees to meet the htlc_minimum while still keeping a lower total fee.
+		if already_collected_value_msat == final_value_msat && payment_paths.len() == 1 {
+			if !hit_minimum_limit {
```
Following tweak doesn't break router unit tests. It might be covered by new test `exact_fee_liquidity_limit()` but our overpaid value reduction logic is making the outcomes similar ?

```
diff --git a/lightning/src/routing/router.rs b/lightning/src/routing/router.rs
index 7f00ccdf..e555575b 100644
--- a/lightning/src/routing/router.rs
+++ b/lightning/src/routing/router.rs
@@ -928,9 +928,9 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
                // by an htlc_minimum_msat value, find another path with a higher value, potentially
                // allowing us to pay fees to meet the htlc_minimum while still keeping a lower total fee.
                if already_collected_value_msat == final_value_msat && payment_paths.len() == 1 {
-                       if !hit_minimum_limit {
-                               break 'paths_collection;
-                       }
+                       //if !hit_minimum_limit {
+                       //      break 'paths_collection;
+                       //}
                        path_value_msat = recommended_value_msat;
                }
        }
```

TheBlueMatt replied:
I believe that line is a strict performance improvement - if we get there the path we found definitely should be the best one, we could do more work, but it won't find a better path.

ariard commented on lightning/src/routing/router.rs:
```diff
@@ -143,12 +143,16 @@ struct RouteGraphNode {
-	value_contribution_msat: u64
+	value_contribution_msat: u64,
+	/// The maximum htlc_minimum_msat along the path, taking into consideration the fees required
+	/// to meet the minimum over the hops required to get there.
```
I think this subtlety around `htlc_minimum_msat` + fees to cover such minimum should be explained better. E.g "The *effective* htlc_minimum_msat required by a hop to forward a payment must also includes fees. `effective_htlc_minimum_msat` = `htlc_minimum_msat` + `fee_base_msat` + (`htlc_minimum_msat` * `fee_proportional_millionths`)

TheBlueMatt commented:
Will respond to comments soon, but there are further issues I need to debug.

TheBlueMatt commented:
Punting to 0.0.14, which we can hopefully accelerate.

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -593,20 +650,12 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
+						// the fees included in $incl_fee_next_hops_htlc_minimum_msat, but also
+						// can't use something that may decrease on future hops.
+						let old_cost = cmp::max(old_entry.total_fee_msat, old_entry.path_htlc_minimum_msat);
+						let new_cost = cmp::max(total_fee_msat, $incl_fee_next_hops_htlc_minimum_msat);
```
This is buggy - `old_entry.path_htlc_minimum_msat` is the `path_htlc_minimum_msat` to the same node on the previous iteration, so needs to match here. probably use the path_htlc instead, plus below.
```suggestion
						let new_cost = cmp::max(total_fee_msat, path_htlc_minimum_msat);
```

TheBlueMatt commented:
Addressed all the bugs I came across, plus some freebie performance tweaks. Still need to address the above comments and add more comments in the changes + write one new test.

TheBlueMatt commented on fuzz/src/router.rs:
```diff
@@ -159,9 +167,11 @@ pub fn do_test<Out: test_logger::Output>(data: &[u8], out: Out) {
+	let mut node_pks: HashSet<_, NonRandomHash> = HashSet::default();
 	let mut scid = 42;
 
+	let mut channel_limits = HashMap::new();
```
Apparently this isn't in the diff anymore? Not sure what it was a reference to.

TheBlueMatt commented:
Ok, only a month later, I believe I've addressed all outstanding feedback on this PR, addressed all the issues I've found, and have at least some marginal tests to detect any regressions in the specific issues I've fixed.

arik-so commented on lightning/src/routing/router.rs:
```diff
@@ -515,18 +579,27 @@
+						let path_htlc_minimum_msat = match compute_fees($incl_fee_next_hops_htlc_minimum_msat, $directional_info.fees)
+								.map(|fee_msat| fee_msat.checked_add($incl_fee_next_hops_htlc_minimum_msat)) {
+							Some(Some(value_msat)) => cmp::max(value_msat, $directional_info.htlc_minimum_msat),
+							_ => u64::max_value()
```
if there is no minimum value, why is the default assumption max_u64, as opposed to 0? Does that imply some hops along the way are misconfigured and hence the route must not be used?

arik-so replied:
Never mind, it's checked_add 🙃

arik-so commented on lightning/src/routing/router.rs:
```diff
@@ -558,120 +628,152 @@
+						{
+							// In test/fuzzing builds, we do extra checks to make sure the skipping
+							// of already-seen nodes only happens in cases we expect (see below).
+							if !should_process { should_process = true; }
```
why wrap it in an if clause?

TheBlueMatt replied:
Because otherwise rust complains that the initial assignment is never read from.

arik-so commented on lightning/src/routing/router.rs:
```diff
@@ -414,25 +459,44 @@
+	let empty_channel_features = ChannelFeatures::empty();
+
+	let mut targets = BinaryHeap::new(); //TODO: Do we care about switching to eg Fibbonaci heap?
+	let mut dist = HashMap::with_capacity(network.get_nodes().len());
```
I think comments for targets and dist would be very appropriate

arik-so commented on lightning/src/routing/router.rs:
`already_processed` or `had_been_processed` might be a bit better than `v`

arik-so commented on lightning/src/routing/router.rs:
afaict, features is only used if skip_node is false? You might be able to combine the skip_node checks from the if clauses above and below and only access features inside it

TheBlueMatt replied:
Yea, doesn't matter much, but, sure.

arik-so commented on lightning/src/routing/router.rs:
```diff
@@ -851,18 +954,29 @@
+					// path.
+					let victim_liquidity = bookkeeped_channels_liquidity_available_msat.get_mut(
+						&payment_path.hops[(payment_path.hops.len() - 1) / 2].0.short_channel_id).unwrap();
+					*victim_liquidity = 0;
```
lol at the name

arik-so commented on lightning/src/routing/router.rs:
```diff
@@ -3400,6 +3538,361 @@
+	fn min_criteria_consistency() {
+		// Test that we don't use an inconsistent metric between updating and walking nodes during
+		// our Dijkstra's pass. In the initial version of MPP, the "best source" for a given node
+		// was updated with a different critera from the heap sorting, resulting in loops in
```
nit: a different criterion

arik-so commented on lightning/src/routing/router.rs:
```diff
@@ -3400,6 +3538,361 @@
+
+		// We construct a network that looks like this:
+		//
+		//            node2 -1(3)2- node3
```
is the bracketed number the value locked up in HTLCs?

TheBlueMatt replied:
Its the short_channel_id. I think this is documented or at least more clear in the graphs in some of the earlier tests.

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -193,20 +202,33 @@
-struct PaymentPath {
-	hops: Vec<PathBuildingHop>,
+struct PaymentPath<'a> {
+	hops: Vec<(PathBuildingHop<'a>, NodeFeatures)>,
```
PathBuildingHop is created for each node as we walk the graph, but PaymentPath only at the end. By avoiding keeping Features (which are Vecs under the hood) in the graph map during the main graph walk, we avoid a lot of Vec clones.

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -414,25 +459,44 @@
 	// This would help to make a better path finding decisions and not "overbook" channels.
 	// It is unaware of the directions (except for `outbound_capacity_msat` in `first_hops`).
-	let mut bookkeeped_channels_liquidity_available_msat = HashMap::new();
+	let mut bookkeeped_channels_liquidity_available_msat = HashMap::with_capacity(network.get_nodes().len());
```
Heh, not my variable name, in transliterated english it makes sense, though :p.

ariard commented on lightning/src/routing/router.rs:
```diff
@@ -193,20 +202,33 @@ struct PathBuildingHop {
+	#[cfg(test)]
+	// In tests, we apply further sanity checks on cases where we skip nodes we already processed
+	// to ensure it is specifically in cases where the fee has gone down because of a decrease in
+	// valud_contribution_msat, which requires tracking it here.
```
"we skip nodes we already processed to ensure it is specifically" which requirement is ensured ? Also "valued_contribution_msat"

ariard commented on lightning/src/routing/router.rs:
```diff
@@ -226,16 +248,14 @@ impl PaymentPath {
-	// and the exception will be triggered.
+	// Sometimes we call this function right after constructing a path which is inconsistent in
+	// that it the value being transferred has decreased while we were doing path finding, leading
+	// to the fees being paid not linig up with the actual limits.
```
nit: "lining up with the actual limits"

ariard commented on lightning/src/routing/router.rs:
```diff
@@ -372,8 +392,42 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
+	//
+	// One potentially problematic case for this algorithm would be if there are many
+	// liquidity-limited paths which are liquidity-limited near the destination (ie early in our
+	// graph walking), we may never find a liquidity-unlimited path which has lower proportional
```
what's a liquidity-unlimited path ? I think it's comparative there and not binary ?

TheBlueMatt replied:
Canged to "a path which is not liquidity-limited".

ariard reviewed (commented):
Ongoing review, just few nits

ariard commented on lightning/src/routing/router.rs:
```diff
@@ -558,140 +634,173 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
+										// different path to the source node that is more expensive,
+										// but which we consider to be cheaper because we are capacity
+										// constrained and the relative fee becomes lower.
+										match compute_fees(minimal_value_contribution_msat, old_entry.src_lowest_inbound_fees)
```
Maybe comment can be clearer "...but on the minimum value that we'll accept flowing over it. Minimal value contribution being a constant, the compared fees from different path to a source node will have the same basis. Thus avoiding to bias toward a capacity-constrained path with a higher relative fee" ?

ariard reviewed (commented):
Ongoing review so far up to dc01b0e

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -372,8 +392,42 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
+	// plus the minimum per-HTLC fee to get from it to another node (aka "shitty pseudo-A*").
+	//
+	// We are not a faithful Dijkstra's implementation because we can change values which impact
+	// earlier nodes while processing later nodes. Specifically, if we reach a channel with a lower
```
Hmm, I read it as "the liquidity limit is lower" not "the lowest liquidity value". I could rephrase to "lower liquidity upper-bound"?

ariard replied:
Yes I like "lower liquidity upper-bound".

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -372,8 +392,42 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
+	// fee (and only lower absolute fee when considering the ultimate value sent). Because we only
+	// consider paths with at least 5% of the total value being sent, the damage from such a case
+	// should be limited, however this could be further reduced in the future by calculating fees
+	// on the amount we wish to route over a path, not the amount we are routing over a path.
```
I changed the sentence to end with "ie ignoring the liquidity limits for the purposes of fee calculation."

TheBlueMatt commented on lightning/src/routing/router.rs:
```diff
@@ -851,18 +961,29 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
+				if !prevented_redundant_path_selection {
+					// If we weren't capped by hitting a liquidity limit on a channel in the path,
+					// we'll probably end up picking the same path again on the next iteration.
+					// Randomly decrease the available liquidity of a hop in the middle of the
```
The issue here is you may be sending 1/1000th of the value of the node in the middle. Without this commit, we just select the same route over and over again and never find any other options. Eventually we should do something somewhat smarter, like checking if a route is exactly identical to one we already selected, but just avoiding the same middle hop solves the issue as well for now.

I dropped the "random", though it was a reference to the somewhat random decision to do this.

ariard replied:
Yeah good to me for now. I don't have a better mitigation against repeated path selection to propose rn.

ariard commented on lightning/src/routing/router.rs:
```diff
@@ -414,25 +459,50 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
 	// This would help to make a better path finding decisions and not "overbook" channels.
 	// It is unaware of the directions (except for `outbound_capacity_msat` in `first_hops`).
-	let mut bookkeeped_channels_liquidity_available_msat = HashMap::new();
+	let mut bookkeeped_channels_liquidity_available_msat = HashMap::with_capacity(network.get_nodes().len());
```
Pre-allocating to the size of the whole graph assumes we'll completely traverse it before reaching `recommended_value_msat` or `path_value_msat` liquidity threshold ? Or what's your thinking there ?

TheBlueMatt replied:
Yea, I mean I guess we could divide by some factor, but why bother, its short-lived and at worst we will, so we might as well just allocate for it.

ariard reviewed (commented):
I think aa9c2c2 is worthy of better explanations. Otherwise, it sounds good to me modulo squashing.

TheBlueMatt commented on lightning/src/routing/router.rs:
I added a note (see diff below) and squashed.

TheBlueMatt commented:
Expanded comment and included some extra fuzzing tests. Diff is
```
$ git diff-tree -U2 1bde367 HEAD
diff --git a/lightning/src/routing/router.rs b/lightning/src/routing/router.rs
index 68815ff2..a74f0cf5 100644
--- a/lightning/src/routing/router.rs
+++ b/lightning/src/routing/router.rs
@@ -212,5 +212,5 @@ struct PathBuildingHop<'a> {
 	/// avoid processing them again.
 	was_processed: bool,
-	#[cfg(test)]
+	#[cfg(any(test, feature = "fuzztarget"))]
 	// In tests, we apply further sanity checks on cases where we skip nodes we already processed
 	// to ensure it is specifically in cases where the fee has gone down because of a decrease in
@@ -637,5 +637,5 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
 								path_htlc_minimum_msat,
 								was_processed: false,
-								#[cfg(test)]
+								#[cfg(any(test, feature = "fuzztarget"))]
 								value_contribution_msat,
 							}
@@ -727,10 +727,10 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
 								old_entry.htlc_minimum_msat = $directional_info.htlc_minimum_msat;
 								old_entry.path_htlc_minimum_msat = path_htlc_minimum_msat;
-								#[cfg(test)]
+								#[cfg(any(test, feature = "fuzztarget"))]
 								{
 									old_entry.value_contribution_msat = value_contribution_msat;
 								}
 							} else if old_entry.was_processed && new_cost < old_cost {
-								#[cfg(test)]
+								#[cfg(any(test, feature = "fuzztarget"))]
 								{
 									// If we're skipping processing a node which was previously
@@ -739,4 +739,18 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
 									// found (which we are processing now) has a lower value
 									// contribution due to an HTLC minimum limit.
+									//
+									// e.g. take a graph with two paths from node 1 to node 2, one
+									// through channel A, and one through channel B. Channel A and
+									// B are both in the to-process heap, with their scores set by
+									// a higher htlc_minimum than fee.
+									// Channel A is processed first, and the channels onwards from
+									// node 1 are added to the to-process heap. Thereafter, we pop
+									// Channel B off of the heap, note that it has a much more
+									// restrictive htlc_maximum_msat, and recalculate the fees for
+									// all of node 1's channels using the new, reduced, amount.
+									//
+									// This would be bogus - we'd be selecting a higher-fee path
+									// with a lower htlc_maximum_msat instead of the one we'd
+									// already decided to use.
 									debug_assert!(path_htlc_minimum_msat < old_entry.path_htlc_minimum_msat);
 									debug_assert!(value_contribution_msat < old_entry.value_contribution_msat);
```

ariard commented on lightning/src/routing/router.rs:
Straightforward comment thanks! Should this say "htlc_minimum" and under ?

TheBlueMatt replied:
Hmm, I dont think so, the issue appears when the `value_contribution_msat` is lower (which is often just the path `htlc_maximum_msat`). The `htlc_minimum_msat` check here is about the criteria for when it should be possible to hit this case.

ariard reviewed (commented):
Code Review ACK 5f5bc41 modulo nit

TheBlueMatt commented:
Gonna merge cause I'm tired of CI failing. I promise to open a new PR to fix the comment if I'm wrong :).
